transaction/std/src/memo/authenticated_sender.rs

```diff
@@ -0,0 +1,197 @@
+    ///
+    /// First, the client should look up the sender's Public Address from their
+    /// hash. If it isn't a known contact we won't be able to authenticate
+    /// them.
+    ///
+    /// Then they need to get the view private key corresponding to the
+    /// subaddress that this TxOut was sent to. This is usually our default
+    /// subaddress view private key.
+    ///
+    /// Finally we can validate the memo against these data. The
+    /// tx_out_public_key is also under the mac, which prevents replay
+    /// attacks.
+    ///
+    /// Arguments:
+    /// * sender_address: The public address of the sender. This can be looked
+    ///   up by the AddressHash provided.
+    /// * recieving_subaddress_view_private_key: This is usually our
+    ///   default_subaddress_view_private_key, but should correspond to whatever
+    ///   subaddress recieved this TxOut.
+    /// * tx_out_public_key: The public key of the TxOut to which this memo is
+    ///   attached.
+    ///
+    /// Returns:
+    /// * subtle::Choice(1u8) if validation passed, subtle::Choice(0u8) if hmac
+    ///   comparison failed.
+    ///
+    /// This function is constant-time.
+    pub fn validate(
+        &self,
+        sender_address: &PublicAddress,
+        recieving_subaddress_view_private_key: &RistrettoPrivate,
+        tx_out_public_key: &CompressedRistrettoPublic,
+    ) -> Choice {
+        let mut result = Choice::from(1u8);
+        let sender_address_hash = AddressHash::from(sender_address);
+        result &= sender_address_hash.ct_eq(&self.sender_address_hash);
+
+        use mc_crypto_keys::KexReusablePrivate;
+        let shared_secret =
+            recieving_subaddress_view_private_key.key_exchange(sender_address.spend_public_key());
+
+        let expected_hmac = Self::compute_hmac(shared_secret.as_ref(), tx_out_public_key);
+        result &= expected_hmac.ct_eq(&self.hmac_value);
+        result
+    }
+
+    // Encapsulates hmac-sha512 computation
+    fn compute_hmac(
+        shared_secret: &[u8; 32],
+        tx_out_public_key: &CompressedRistrettoPublic,
+    ) -> [u8; 16] {
+        let mut mac = HmacSha512::new_from_slice(shared_secret.as_ref())
+            .expect("hmac can take a key of any size");
+        mac.update(tx_out_public_key.as_ref());
+        let mut result = [0u8; 16];
+        result.copy_from_slice(&mac.finalize().into_bytes()[0..16]);
+        result
+    }
+}
+
+impl From<&[u8; 32]> for AuthenticatedSenderMemo {
+    fn from(src: &[u8; 32]) -> Self {
+        let address_hash: [u8; 16] = src[0..16].try_into().expect("arithmetic error");
+        let hmac_value: [u8; 16] = src[16..32].try_into().expect("arithmetic error");
+        Self {
+            sender_address_hash: address_hash.into(),
+            hmac_value,
+        }
+    }
+}
+
+impl From<AuthenticatedSenderMemo> for MemoPayload {
+    fn from(src: AuthenticatedSenderMemo) -> MemoPayload {
+        let memo_type = [0u8, 1u8];
```

Comment:
Should probably be in a list of constants somewhere to make it easy to see which types we are reserving for ourselves in this crate